web/contribute.rst

```diff
@@ -1,230 +1,458 @@
 
-The scikit-bio coding guidelines describe our :doc:`expectations for unit tests <devdoc/code_guide>`. You should review the unit test section before working on your test code.
+Functional test
+^^^^^^^^^^^^^^^
 
-Tests can be executed by running ``make test`` from the base directory of the project or from within a Python or IPython session::
+You will want to test your code like a *user* would do: import the function, execute it on some input data, and examine whether the results are correct. You may install additional software such as `Jupyter <https://jupyter.org/>`_ in the same conda environment to make the testing process convenient and pleasant. There is no need to reinstall the modified scikit-bio package in a separate environment in order to test. As soon as you edit the code, the changes will immediately reflect when you use the code.
+
+For example, you wrote a function ``gc_content``, which calculates the fraction of ``G`` and ``C`` in a nucleotide sequence::
+
+    def gc_content(seq):
+        return sum(1 for x in seq if x in "GC") / len(seq)
+
+You added this function to the code file ``skbio/sequence/nucl.py``. It will be available for use in any Python code launched in the same conda environment::
+
+    >>> from skbio.sequence.nucl import gc_content
+
+Now test the function on some data. For example, you would expect that ``gc_content("ACGT")`` and ``gc_content("GGATCCGC")`` return 0.5 and 0.75, respectively. Is that the case?
+
+We highly recommend that you use real-world biological data in addition to small, dummy data for testing. This will let you evaluate the robustness and scalability of your code in real-world applications. For example, DNA sequences retrieved from public databases may contain lowercase letters, and you will find that the ``gc_content`` function cannot handle them properly. For another instance, a FASTQ file with ten million sequences (which is common) may cost a function forever to process, in which case you should consider optimization.
+
+Unit test
+^^^^^^^^^
+
+`Unit testing <https://en.wikipedia.org/wiki/Unit_testing>`_ involves testing the smallest units of your code, such as classes, functions, and methods, to ensure they function correctly in isolation. It is a fundamental best practice in software engineering, but is often overlooked by beginners. Unit testing is automated by writing test code alongside the algorithm code. Both types of code are integrated into the scikit-bio codebase. This test code is then regularly executed whenever changes are made to ensure that the intended behavior remains consistent over time.
```

Comment:
"Unit testing is made easier by writing test code alongside the algorithm code." may be more accurate than "Unit testing is automated by writing test code alongside the algorithm code."